framework/src/loops/ComputeUserObjectsThread.C

```diff
@@ -191,6 +198,40 @@ ComputeUserObjectsThread::onInternalSide(const Elem * elem, unsigned int side)
   }
 }
 
+void
+ComputeUserObjectsThread::onInterface(const Elem * elem, unsigned int side, BoundaryID bnd_id)
+{
+  // Pointer to the neighbor we are currently working on.
+  const Elem * neighbor = elem->neighbor_ptr(side);
+
+  // Get the global id of the element and the neighbor
+  const dof_id_type elem_id = elem->id(), neighbor_id = neighbor->id();
+
+  if (!_interface_user_objects.hasActiveBoundaryObjects(bnd_id, _tid))
+    return;
+
+  if (!((neighbor->active() && (neighbor->level() == elem->level()) && (elem_id < neighbor_id)) ||
+        (neighbor->level() < elem->level())))
+    return;
+
+  _fe_problem.prepareFace(elem, _tid);
+  _fe_problem.reinitNeighbor(elem, side, _tid);
+
+  // Set up Sentinels so that, even if one of the reinitMaterialsXXX() calls throws, we
+  // still remember to swap back during stack unwinding.
+  SwapBackSentinel face_sentinel(_fe_problem, &FEProblem::swapBackMaterialsFace, _tid);
+  _fe_problem.reinitMaterialsFace(elem->subdomain_id(), _tid);
+
+  SwapBackSentinel neighbor_sentinel(_fe_problem, &FEProblem::swapBackMaterialsNeighbor, _tid);
+  _fe_problem.reinitMaterialsNeighbor(neighbor->subdomain_id(), _tid);
+
+  const auto & objects = _interface_user_objects.getActiveBoundaryObjects(bnd_id, _tid);
+  for (const auto & uo : objects)
+    uo->execute();
+
+  _fe_problem.reinitMaterialsBoundary(bnd_id, _tid);
```

Comment:
This is unfortunately something that works for your particular use case, e.g. compute face and neighbor material properties, operate on them in the user object, and then pass them into a boundary restricted material. But what if someone wanted to create an `InterfaceUserObject` that has boundary restricted material properties as a pre-requisite instead of the reverse? Our precedent is to evaluate material properties first and then evaluate user objects, and I think we should stick to it. Hence our belief that we should add `InterfaceMaterial`